File: backend/app/schemas/verification.py
"""
IntelliHall — Verification Pydantic Schemas

Defines request/response shapes for the Hall Verification workflow.

    VerificationRequestRead   – Student card shown to the Hall Admin
    VerificationActionRequest – Body for approve/reject endpoints
    PaginatedVerificationResponse – Paginated list of verification requests
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import HallVerificationStatus


# ---------------------------------------------------------------------------
# Student card — read-only, shown to the admin
# ---------------------------------------------------------------------------

class VerificationRequestRead(BaseModel):
    """
    Student data shown to the Hall Admin on the Pending Verifications page.
    Mirrors the relevant subset of UserRead — no sensitive fields.
    """

    model_config = ConfigDict(from_attributes=True)

    id: str = Field(..., description="User UUID.")
    name: str = Field(..., description="Student's full name.")
    email: str = Field(..., description="Student's e-mail address.")
    roll_number: str | None = Field(default=None, description="Student roll number.")
    room_number: str | None = Field(default=None, description="Requested room number.")
    hall_id: str | None = Field(default=None, description="Requested hall UUID.")
    hall_name: str | None = Field(default=None, description="Requested hall name.")
    hall_verification_status: HallVerificationStatus = Field(
        ..., description="Current verification state."
    )
    hall_rejection_reason: str | None = Field(
        default=None, description="Rejection reason (if applicable)."
    )
    created_at: datetime = Field(..., description="When the account was created.")
    updated_at: datetime = Field(..., description="When the record was last updated.")


# ---------------------------------------------------------------------------
# Admin action — approve or reject a student
# ---------------------------------------------------------------------------

class VerificationActionRequest(BaseModel):
    """Body for POST /verification/{user_id}/approve and /reject."""

    rejection_reason: str | None = Field(
        default=None,
        max_length=1000,
        description="Optional reason text when rejecting. Ignored on approve.",
    )


# ---------------------------------------------------------------------------
# Paginated response wrapper
# ---------------------------------------------------------------------------

class PaginatedVerificationResponse(BaseModel):
    """Paginated list of verification requests."""

    items: list[VerificationRequestRead]
    total: int = Field(..., description="Total matching records.")
    page: int = Field(..., description="Current page number (1-indexed).")
    page_size: int = Field(..., description="Number of items per page.")
    total_pages: int = Field(..., description="Total number of pages.")

File: frontend/src/services/verification.ts
import { apiClient } from "@/services/api-client";
import type {
  PaginatedVerificationResponse,
  UpdateHallRequest,
  User,
  VerificationRequestRead,
} from "@/types/auth";

// ---------------------------------------------------------------------------
// GET /api/v1/verification/pending
// ---------------------------------------------------------------------------

export async function getPendingVerifications(
  params?: { page?: number; page_size?: number }
): Promise<PaginatedVerificationResponse> {
  const response = await apiClient.get<PaginatedVerificationResponse>(
    "/verification/pending",
    { params }
  );
  return response.data;
}

// ---------------------------------------------------------------------------
// POST /api/v1/verification/{user_id}/approve
// ---------------------------------------------------------------------------

export async function approveVerification(userId: string): Promise<VerificationRequestRead> {
  const response = await apiClient.post<VerificationRequestRead>(
    `/verification/${userId}/approve`
  );
  return response.data;
}

// ---------------------------------------------------------------------------
// POST /api/v1/verification/{user_id}/reject
// ---------------------------------------------------------------------------

export async function rejectVerification(
  userId: string,
  rejectionReason?: string | null
): Promise<VerificationRequestRead> {
  const response = await apiClient.post<VerificationRequestRead>(
    `/verification/${userId}/reject`,
    { rejection_reason: rejectionReason ?? null }
  );
  return response.data;
}

// ---------------------------------------------------------------------------
// PATCH /api/v1/users/me/hall  (student self-service)
// ---------------------------------------------------------------------------

export async function updateMyHall(data: UpdateHallRequest): Promise<User> {
  const response = await apiClient.patch<User>("/users/me/hall", data);
  return response.data;
}
